### PR TITLE
TSK-1504: Next scheduled job now based on due date of previous job

### PR DESCRIPTION
**Pull request by tge20**
---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [x] Commit message format → TSK-XXX: Your commit message.
- [x] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [x] SonarCloud analysis meets our standards
- [x] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [x] PR fulfills the ticket
- [x] Edge cases and unwanted side effects are tested
- [x] Readability

### CODE REVIEW

**Comment by tge20** [comment]
https://sonarcloud.io/dashboard?branch=TSK-1504&id=tge20_taskana

File: lib/taskana-core/src/main/java/pro/taskana/common/internal/jobs/AbstractTaskanaJob.java [hunk 1]
```diff
@@ -54,4 +58,17 @@ public static TaskanaJob createFromScheduledJob(
...
+
+  protected Instant getNextDueForCleanupJob() {
+    Instant nextRun = Instant.now();
+    if (scheduledJob != null && scheduledJob.getDue() != null) {
+      nextRun = scheduledJob.getDue();
+    }
+
+    while (nextRun.isBefore(Instant.now())) {
+      nextRun = nextRun.plus(runEvery);
+    }
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/common/internal/jobs/AbstractTaskanaJob.java:70]: Why did you wrap this in a while loop? 
Isn't a simple addition enough?

File: lib/taskana-core/src/test/java/acceptance/jobs/TaskCleanupJobAccTest.java [hunk 2]
```diff
@@ -113,11 +119,40 @@ void should_DeleteOldTaskCleanupJobs_When_InitializingSchedule() throws Exceptio
...
+    assertThat(jobsToRun).hasSize(1);
+    assertThat(jobsToRun.get(0).getDue()).isEqualTo(firstDue);
+
+    JobRunner runner = new JobRunner(taskanaEngine);
+    runner.runJobs();
+    Duration runEvery = taskanaEngineConfiguration.getCleanupJobRunEvery();
+    jobsToRun = jobMapper.findJobsToRun(Instant.now().plus(runEvery));
+
+    assertThat(jobsToRun).hasSize(1);
+    assertThat(jobsToRun.get(0).getDue()).isEqualTo(firstDue.plus(runEvery));
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/jobs/TaskCleanupJobAccTest.java:153]:
```suggestion
    assertThat(jobsToRun).extracting(ScheduledJob::getDue).containsExactly(firstDue.plus(runEvery));
```

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

File: lib/taskana-core/src/main/java/pro/taskana/common/internal/jobs/AbstractTaskanaJob.java (hunk 1 above)
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/common/internal/jobs/AbstractTaskanaJob.java:70]: It is a measure to ABSOLUTELY ensure that the problem does not appear. A possible problem could appear if the due of the previous job is far in the past, for whatever reason. 
I also thought about just adding runEvery to the current time, no matter what the due of the last scheduled job has been.

Comment (mustaphazorgati) [reply lib/taskana-core/src/main/java/pro/taskana/common/internal/jobs/AbstractTaskanaJob.java:70]: > It is a measure to ABSOLUTELY ensure that the problem does not appear. A possible problem could appear if the due of the previous job is far in the past, for whatever reason.
What's the issue here? What will happen when the entry will be set to a date in the past? I am not 100% sure, but I think on the next check that job will just be executed again. Isn't that the behaviour we want? Maybe we can talk with @holgerhagen or @gitgoodjhe about the expected behaviour here.

**Review by mustaphazorgati — APPROVED** [review]

**Review by mustaphazorgati — APPROVED** [review]

**Review by mustaphazorgati — APPROVED** [review]

**Pull request merged**